### PR TITLE
[Feature] Emails & User Verification

### PR DESCRIPTION
**Pull request by Tanax-Xt**
### Description

Added functions for creating and sending emails. Verification of user's email via a JWT token

### Changes

* Update `.env` (and create `.env.example`). Update `.gitignore`
* Update `requirements/common.txt`
* Add `verify_email()` and `verify_user()` for send email and verify user from email in `src/api/v1/auth/routes.py`
* Update `config.py`
* Update `create_access_token()` in `src/core/security.py` - it became possible to pass the first argument not only of the int type and set the token lifetime with the second argument
* Create `decode_access_token()` in `src/core/security.py`
* In `src/email/service.py` create new funcs for generating emails, sending emails, generating emails for verifying user
* Create `src/email/templates/build_email/verify_email.html`

### Related Issues

Complete task https://github.com/orgs/rapid-integration/projects/1/views/1?pane=issue&itemId=70360172

### CODE REVIEW

File: backend/requirements/common.txt [hunk 1]
```diff
@@ -18,10 +19,14 @@ pyheck
...
 
 # Security
 passlib
 bcrypt==4.0.1  # https://github.com/pyca/bcrypt/issues/684
 
 # Async files
 aiofiles
+
+# HTML-templases
+jinja2
```
Comment (zobweyt) [inline backend/requirements/common.txt]: isn't this package already included in `fastapi[all]`?

File: backend/requirements/common.txt [hunk 2]
```diff
@@ -18,10 +19,14 @@ pyheck
...
+jwt
 
 # Security
 passlib
 bcrypt==4.0.1  # https://github.com/pyca/bcrypt/issues/684
 
 # Async files
 aiofiles
+
+# HTML-templases
```
Comment (zobweyt) [inline backend/requirements/common.txt]:
```suggestion
# HTML Templates
```

File: backend/src/api/v1/auth/routes.py [hunk 3]
```diff
@@ -38,3 +40,30 @@ def login(form: OAuth2Form, session: TransactionalSession) -> Any:
 
     access_token = create_access_token(user.id)
     return TokenPayload(access_token=access_token)
+
+
+@router.post("/verify-email")
+async def verify_email(email: str, session: TransactionalSession) -> Any:
+    email = str(email)
```
Comment (zobweyt) [inline backend/src/api/v1/auth/routes.py]: unnecessary type conversion

File: backend/src/emails/templates/build_email/verify_email.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    </style>
+</head>
+<body>
+    <div class="container">
+        <h2>Verify your email address</h2>
+        <p>Hi {{ email }},</p>
+        <p>Please click the button below to verify your email address:</p>
+        <a href="{{ link }}" class="button">Verify Email</a>
+        <p class="link">If the button doesn't work, please <a href="{{ link }}">click here</a></p>
+        <p>This link will expire on {{ expires_at }} minutes.</p>
```
Comment (zobweyt) [inline backend/src/emails/templates/build_email/verify_email.html]:
```suggestion
        <p>This link will expire in {{ expires_at }} minutes.</p>
```

File: backend/src/core/security.py [hunk 5]
```diff
@@ -9,13 +10,21 @@
...
-def create_access_token(subject: int) -> str:
-    expires_at = datetime.now(timezone.utc) + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
+def create_access_token(subject: int | str, expire=settings.ACCESS_TOKEN_EXPIRE_MINUTES) -> str:
+    expires_at = datetime.now(timezone.utc) + timedelta(minutes=expire)
     to_encode = JWTPayload(exp=expires_at, sub=subject)
     encoded_jwt = jwt.encode(to_encode.model_dump(), settings.SECRET_KEY, algorithm=settings.ALGORITHM)
     return encoded_jwt
 
 
+def decode_access_token(token: str):
```
Comment (zobweyt) [inline backend/src/core/security.py]:
```suggestion
def decode_access_token(token: str) -> str | None:
```

File: backend/src/emails/service.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    return message
+
+
+async def send_email(message: MessageSchema, template_name: str):
+    fm = FastMail(settings.MAIL_CONNECTION_CONF)
+    await fm.send_message(message, template_name=template_name)
+    return JSONResponse(status_code=200, content={"message": "email has been sent"})
+
+
+def generate_verify_user_email_email(email_to: str, token: str, expires_at: int):
```
Comment (zobweyt) [inline backend/src/emails/service.py]: rename it to generate_user_verification_email

File: backend/requirements/common.txt [hunk 7]
```diff
@@ -1,5 +1,6 @@
 # Web Framework
 fastapi[all]
+fastapi_mail
```
Comment (zobweyt) [inline backend/requirements/common.txt]: use new notation
```suggestion
fastapi-mail
```

File: backend/src/emails/service.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        recipients=recipients,
+        template_body=template_body,
+        subtype=subtype)
+    return message
+
+
+async def send_email(message: MessageSchema, template_name: str):
+    fm = FastMail(settings.MAIL_CONNECTION_CONF)
+    await fm.send_message(message, template_name=template_name)
+    return JSONResponse(status_code=200, content={"message": "email has been sent"})
```
Comment (zobweyt) [inline backend/src/emails/service.py]: don't return json response in a service function. it's not a part of the API. do it in the endpoint function instead

File: backend/src/emails/service.py (hunk 6 above)
Comment (zobweyt) [inline backend/src/emails/service.py]: add return type hint

File: backend/src/emails/service.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+                   subtype: MessageType = MessageType.html) -> MessageSchema:
+    message = MessageSchema(
+        subject=subject,
+        recipients=recipients,
+        template_body=template_body,
+        subtype=subtype)
+    return message
+
+
+async def send_email(message: MessageSchema, template_name: str):
```
Comment (zobweyt) [inline backend/src/emails/service.py]:
```suggestion
async def send_email(message: MessageSchema, template_name: str) -> None:
```

File: backend/src/emails/service.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
+from fastapi_mail import FastMail, MessageSchema, MessageType
+from starlette.responses import JSONResponse
+
+from src.core.config import settings
+
+
+def generate_email(subject: str, recipients: list, template_body: dict,
+                   subtype: MessageType = MessageType.html) -> MessageSchema:
```
Comment (zobweyt) [inline backend/src/emails/service.py:7]: format arguments wrapping here

File: backend/src/core/security.py [hunk 11]
```diff
@@ -9,13 +10,21 @@
...
+    try:
+        decoded_token = jwt.decode(token, settings.SECRET_KEY, algorithms=settings.ALGORITHM)
```
Comment (zobweyt) [inline backend/src/core/security.py:22]: make it type safe. add this line below: `data = JWTPayload(**decoded_token)`

File: backend/src/core/security.py [hunk 12]
```diff
@@ -9,13 +10,21 @@
...
+        return str(decoded_token["sub"])
```
Comment (zobweyt) [inline backend/src/core/security.py]: and after changing above do this:
```suggestion
        return data.sub
```

File: backend/src/core/config.py [hunk 13]
```diff
@@ -37,5 +39,47 @@ def DATABASE_URI(self) -> PostgresDsn:
...
+            MAIL_STARTTLS=self.MAIL_STARTTLS,
+            MAIL_SSL_TLS=self.MAIL_SSL_TLS,
+            USE_CREDENTIALS=self.USE_CREDENTIALS,
+            VALIDATE_CERTS=self.VALIDATE_CERTS,
+            TEMPLATE_FOLDER=self.TEMPLATE_FOLDER
+        )
+
+    DOMAIN: str = "localhost"
+    PORT: int = 8000
+    ENVIRONMENT: Literal["local", "staging", "production"] = "local"
```
Comment (zobweyt) [inline backend/src/core/config.py]: is it even used somewhere? we use `DEBUG` instead of it

File: backend/src/core/config.py [hunk 14]
```diff
@@ -37,5 +39,47 @@ def DATABASE_URI(self) -> PostgresDsn:
...
+
+    @computed_field  # type: ignore[misc]
```
Comment (zobweyt) [inline backend/src/core/config.py]:
```suggestion
    @computed_field
```

File: backend/src/core/config.py [hunk 15]
```diff
@@ -37,5 +39,47 @@ def DATABASE_URI(self) -> PostgresDsn:
...
+    @property
+    def server_host(self) -> str:
+        # Use HTTPS for anything other than local development
+        if self.ENVIRONMENT == "local":
```
Comment (zobweyt) [inline backend/src/core/config.py]:
```suggestion
        if self.DEBUG:
```

File: backend/src/core/config.py [hunk 16]
```diff
@@ -37,5 +39,47 @@ def DATABASE_URI(self) -> PostgresDsn:
...
+    MAIL_PASSWORD: str
+    MAIL_FROM: str
+    MAIL_FROM_NAME: str
+    MAIL_STARTTLS: bool
+    MAIL_SSL_TLS: bool
+    USE_CREDENTIALS: bool
+    VALIDATE_CERTS: bool
+    MAIL_PORT: int
+    EMAIL_RESET_TOKEN_EXPIRE_MINUTES: int
+    TEMPLATE_FOLDER: str
```
Comment (zobweyt) [inline backend/src/core/config.py]: rename it to `TEMPLATES_PATH`

File: backend/.env.example [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+MAIL_PASSWORD=password
+MAIL_FROM=email@gmail.com
+MAIL_FROM_NAME=Name
+MAIL_STARTTLS=True
+MAIL_SSL_TLS=False
+USE_CREDENTIALS=True
+VALIDATE_CERTS=True
+MAIL_PORT=587
+EMAIL_RESET_TOKEN_EXPIRE_MINUTES=10
+TEMPLATE_FOLDER=src/email/templates/build_email
```
Comment (zobweyt) [inline backend/.env.example]: rename it to `TEMPLATES_PATH`

**Review by zobweyt — CHANGES_REQUESTED** [review]

File: backend/src/core/config.py (hunk 15 above)
Comment (Tanax-Xt) [reply backend/src/core/config.py]: what's the difference?

File: backend/src/core/config.py (hunk 13 above)
Comment (Tanax-Xt) [reply backend/src/core/config.py]: This is used, but DEBUG is not

File: backend/requirements/common.txt (hunk 1 above)
Comment (Tanax-Xt) [reply backend/requirements/common.txt]: no

File: backend/requirements/common.txt (hunk 2 above)
Comment (Tanax-Xt) [reply backend/requirements/common.txt]: no in the last commit

File: backend/src/core/config.py (hunk 13 above)
Comment (zobweyt) [reply backend/src/core/config.py]: where is `ENVIRONMENT` being used? `DEBUG` is used even at `main.py`

File: backend/src/core/config.py (hunk 15 above)
Comment (zobweyt) [reply backend/src/core/config.py]: do not add extra setting (`ENVIRONMENT`). see conversation above

File: backend/src/emails/service.py (hunk 10 above)
Comment (Tanax-Xt) [reply backend/src/emails/service.py:7]: Already so

File: backend/requirements/common.txt (hunk 1 above)
Comment (zobweyt) [reply backend/requirements/common.txt]: https://github.com/tiangolo/fastapi/blob/912524233b535a1d45b54863b2c4e0bd2464b193/pyproject.toml#L84

File: backend/src/core/security.py (hunk 11 above)
Comment (Tanax-Xt) [reply backend/src/core/security.py:22]: I don't understand you

File: backend/requirements/common.txt (hunk 1 above)
Comment (Tanax-Xt) [reply backend/requirements/common.txt]: ok

Comment (zobweyt) [reply backend/requirements/common.txt]: it's not necessary to include it in requirements as it's already included by `fastapi[all]`. see https://github.com/tiangolo/fastapi/blob/912524233b535a1d45b54863b2c4e0bd2464b193/pyproject.toml#L84

File: backend/src/api/v1/auth/routes.py [hunk 18]
```diff
@@ -38,3 +41,29 @@ def login(form: OAuth2Form, session: TransactionalSession) -> Any:
...
+async def send_email_to_verify_user(email_to: str, session: TransactionalSession) -> Any:
+    user = get_user_by_email(session, email_to)
+    if not user:
+        raise HTTPException(status.HTTP_404_NOT_FOUND, "User not found")
+    if user.is_verified:
+        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, "Email already verified")
+    token = create_access_token(email_to, expire=settings.EMAIL_RESET_TOKEN_EXPIRE_MINUTES)
+    message = generate_user_verification_email(email_to, token, settings.EMAIL_RESET_TOKEN_EXPIRE_MINUTES)
+    await send_email(message, "verification-email.html")
+    return JSONResponse(status_code=200, content={"message": "email has been sent"})
```
Comment (zobweyt) [inline backend/src/api/v1/auth/routes.py]:
```suggestion
    return {"detail": "Email has been sent"}
```

**Review by zobweyt — CHANGES_REQUESTED** [review]

File: backend/requirements/common.txt (hunk 1 above)
Comment (Tanax-Xt) [reply backend/requirements/common.txt]: done

**Review by zobweyt — APPROVED** [review]

**Pull request merged**